### PR TITLE
Update/clarify required properties for prefix extractors

### PR DESCRIPTION
**Pull request by pdillinger**
Summary: Most of the properties listed as required for prefix extractors
are not really required but offer some conveniences. This updates API
comments to clarify actual requirements, and adds tests to demonstrate
how previously presumed requirements can be safely violated.

This might seem like a useless exercise, but this relaxing of requirements
would be needed if we generalize prefixes to group keys not just at the
byte level but also based on bits or arbitrary value ranges. For
applications without a "natural" prefix size, having only byte-level
granularity often means one prefix size to the next differs in magnitude
by a factor of 256.

Test Plan: Tests added, also covering missing Iterator cases from #10244

### CODE REVIEW

File: db/db_bloom_filter_test.cc [hunk 1]
```diff
@@ -3113,6 +3113,382 @@ TEST_F(DBBloomFilterTest, SeekForPrevWithPartitionedFilters) {
...
+  options.table_factory.reset(NewBlockBasedTableFactory(bbto));
+  options.memtable_prefix_bloom_size_ratio = 0.1;
+  options.statistics = CreateDBStatistics();
+
+  DestroyAndReopen(options);
+
+  ASSERT_OK(Put("aaaa123", "val1"));
+  ASSERT_OK(Put("aaaa211", "val2"));
+  ASSERT_OK(Put("aaaa900", "val3"));
+  ASSERT_OK(Put("aab", "val4"));  // out of domain
```
Comment (bjlemaire) [inline db/db_bloom_filter_test.cc:3263]: Somewhat unrelated to this PR: if I understand correctly any key shorter than the prefix extractor length escapes any bloom filter check.  
Would it be worthwhile to consider a padding option for the keys shorter than the prefix extractor length to integrate them into the Bloom filter (at least as an option to the user, the default padding char could be the null byte "\0") ? E.g: "aab" could be added to the BF as "aab\0". This adds a small chance of false positive if the actual key "aab\0" is queried, but if the length of the prefix extractor is large, a lot of keys may fall out of the domain of the BF.

Comment (bjlemaire) [reply db/db_bloom_filter_test.cc:3263]: (Note: if the padding actually becomes an option, note that the axiom `key.starts_with(prefix(key, padding=true)) == true` is no longer verified)

Comment (pdillinger) [reply db/db_bloom_filter_test.cc:3263]: The challenge with this is performance. Computing the right hash requires either materializing a copy of the key with the extra data or operating the hash function in a streaming mode, which is inefficient for a small key (and requires a streaming implementation of the hash function).

Comment (pdillinger) [reply db/db_bloom_filter_test.cc:3263]: To have the same effect you could use a prefix extractor that omits trailing 0 bytes. This is actually a pretty good idea for a prefix extractor that has some of the better properties of both fixed and capped. Like fixed, all keys are in domain. Like capped, there are exactly 256^n possible prefixes.

File: db/db_bloom_filter_test.cc [hunk 2]
```diff
@@ -3113,6 +3113,382 @@ TEST_F(DBBloomFilterTest, SeekForPrevWithPartitionedFilters) {
...
+class BackwardBytewiseComparator : public Comparator {
+ public:
+  const char* Name() const override { return "BackwardBytewiseComparator"; }
+
+  int Compare(const Slice& a, const Slice& b) const override {
+    size_t min_size = std::min(a.size(), b.size());
+    const char* a_end = a.data() + a.size();
+    const char* b_end = b.data() + b.size();
+    for (size_t i = 1; i <= min_size; ++i) {
+      if (a_end[-i] == b_end[-i]) {
```
Comment (bjlemaire) [inline db/db_bloom_filter_test.cc:3126]: `i` is unsigned (`size_t`) => `-i` is also unsigned ("-1" becomes 128, or 255 I believe)

File: db/db_bloom_filter_test.cc (hunk 1 above)
Comment (pdillinger) [reply db/db_bloom_filter_test.cc:3263]: > if I understand correctly any key shorter than the prefix extractor length escapes any bloom filter check.

The difference between fixed and capped is that fixed excludes them from the check as you describe and capped considers them to be smaller prefixes.

**Pull request closed without merging**